Tree Borrows without protector (Path Provenance) (#220)
minirust/minirust: spec/mem/tree_borrows/tree.md

# Tree Structure in Tree Borrows

> The core data structure of the Tree Borrows is a *Tree*. We use a tree to track reborrows in each allocation. The tree consists of several nodes, each corresponding to a borrow tag (i.e., a reborrow operation). Each node has a parent and a list of children. Additionally, on memory locations (bytes) within the allocation, we track each tag's state defined in [state_machine.md](state_machine.md).

Reviewer: ```suggestion
The core data structure of Tree Borrows is a *tree*, with a state machine in each node.
We use a tree to track reborrows in each allocation; each reborrow adds a new node to the tree.
The per-node state machine is defined in [state_machine.md](state_machine.md).

Structurally, we use the usual function representation of a tree: we store a list of children.
```

```rust
struct Node {
    children: List<Node>,
    /// State for each location
    location_states: List<LocationState>,
}
```

> During each memory access, we update states according to the state machine defined in [state_machine.md](state_machine.md). When a node is accessed, each node in the tree can be divided into two disjoint sets: *Child* and *Foreign*, based on its relative position to the accessed node. The Child set includes the node itself and all its descendants, while the Foreign set contains all other nodes.

Reviewer: ```suggestion
During each memory access, we update states according to the state machine.
When a node is accessed, each node in the tree can be divided into two disjoint sets: *child nodes* and *foreign nodes*, based on its relative position to the accessed node.
The *child* set includes the node itself and all its descendants, while the *foreign* set contains all other nodes.
```

```rust
enum NodeRelation {
    Child, 
    Foreign,
}
```

The state transition depends on both the node relation and the type of access operation.

```rust
enum AccessKind {
    Read, 
    Write,
}
```

Then we implement how to actually update and check states in a Tree during each memory access.

```rust
impl Node {
    /// Perform state transition on all locations of `self`.
    /// The `node_relation` indicates whether the accessed node is a child or foreign
    /// *from the perspective of the `self`*.
    fn transition(
        &mut self, 
        node_relation: NodeRelation,
        access_kind: AccessKind,
        offset_in_alloc: Size,
        size: Size
    ) -> Result {
        let offset_start = offset_in_alloc.bytes();
        for offset in offset_start..offset_start + size.bytes() {

>             let curr_state = self.location_states[offset];

Reviewer: ```suggestion
            // FIXME: Use `try_mutate_at` once that works with `Result`.
            let curr_state = self.location_states[offset];
```

            let next_state = curr_state.transition(access_kind, node_relation)?;
            self.location_states.set(offset, next_state);
        }

        ret(())
    }

    /// Recusively do state transition on the `self` node and all its descendants.

>     /// The `path` is the path from the `self` to the accessed node.
>     /// The `is_ancestor` indicates whether the `self` is an ancestor of the accessed node.

Reviewer: ```suggestion
    /// `path` is the path from the `self` to the accessed node.
    /// `is_ancestor` indicates whether `self` is an ancestor of the accessed node.
```

    ///
    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
    fn access(
        &mut self,

>         path: Path, // root -> child

Reviewer: ```suggestion
        path: Path, // self -> child
```
I think?

        access_kind: AccessKind,
        offset_in_alloc: Size,
        size: Size,

>         is_ancestor: bool,

Reviewer: I am confused by this, it seems redundant with `path`? Specifically, if there is a path from `self` to the accessed node, then `self` must be an ancestor of the accessed node?

Would it work to have an `Option<Path>`, where `None` means "accessed node is not a descendant of `self`"?

    ) -> Result {
        // Indicates whether the accessed node is a child or foreign

>         // *from the perspective of the `self`*.

Reviewer: ```suggestion
        // *from the perspective of `self`*.
```
There should be never a `the` before `self`.

        //
        // If the `self` is an ancestor of the accessed node, the accessed node is a child of the `self`.
        // Otherwise, the accessed node is a `foreign` of the `self`
        let node_relation = if is_ancestor { NodeRelation::Child } else { NodeRelation::Foreign };

        // perform state transition on the `self`
        self.transition(node_relation, access_kind, offset_in_alloc, size)?;

        let children_num = self.children.len();

>         // If the `self` is leaf, stop here.

Reviewer: ```suggestion
        // If `self` is a leaf, stop here.
```

        if children_num == Int::ZERO { return ret(()); }

>         // The path from the sub-root

Reviewer: Here things become somewhat more complicated than I expected.

What about the pseudo-code?
* for all children (so no `if children_num == Int::ZERO` needed, the loop will just never run then)
  * if the path starts with this child, do a child access with the path shortened by the first element
  * otherwise, do a child access with the path being `None` (i.e., child is not an ancestor of accessed node)

        let sub_path = if path.len() == Int::ZERO { path } else { path.subslice_with_length(Int::ONE, path.len() - Int::ONE) };

        let mut ancestor_mask = list![false; children_num];

        // If the `self` is not an ancestor of the accessed node,
        // no child of the `self` could be an ancestor of the accessed node
        if is_ancestor {
            if let Some(sub_root_id) = path.first() { ancestor_mask.set(sub_root_id, is_ancestor); }
        }

        self.children = self.children.zip(ancestor_mask).try_map(|(mut child, child_is_ancestor)| {
            child.access(sub_path, access_kind, offset_in_alloc, size, child_is_ancestor)?;
            ret::<Result<Node>>(child)
        })?;

        ret(())
    }
}
```

We also implement some helper methods for manipulating Trees.

```rust
impl Node {
    /// Add a new child node to the tree whose root is self

>     /// The `path` is the path from the `self` to the parent of the `node`.

Reviewer: ```suggestion
    /// `path` is the path from `self` to the parent of `node`.
```

    ///
    /// Return the path from the `self` to the `node`
    fn add_node(&mut self, parent_path: Path, child: Node) -> Result<Path> {

>         if parent_path.len() == Int::ZERO {

Reviewer: ```suggestion
        // If this is where we want to insert, we are done.
        if parent_path.len() == Int::ZERO {
```

>             self.children.push(child);
>             return ret(list![self.children.len() - Int::ONE]);

Reviewer: ```suggestion
            let child_idx = self.children.len();
            self.children.push(child);
            return ret(list![child_idx]);
```

        }

>         // If the `self` is leaf, we fail to find the parent node.

Reviewer: ```suggestion
        // If `self` is a leaf and the path keeps going, then the path is invalid.
```

        if self.children.len() == Int::ZERO {

>             panic!("Tree Borrows: Not found the parent node");

Reviewer: ```suggestion
            panic!("Node::add_node: invalid parent path");
```

        }

        // The `path` indicates which node to access next; call it the sub-root.
        let sub_root_id = parent_path.first().unwrap();

        // The path from the sub-root to the child

>         let sub_path = parent_path.subslice_with_length(Int::ONE, parent_path.len() - Int::ONE);

Reviewer: We should add a `split_first` method to `List`, like [this one](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.split_first).

>         self.children.mutate_at(sub_root_id, |sub_root| {

Reviewer: ```suggestion
        // Find the right child, and add the node there.
        self.children.mutate_at(sub_root_id, |sub_root| {
```

>             let mut child_path = sub_root.add_node(sub_path, child)?;
>             // FIXME: use push_front

Reviewer: ```suggestion
            let mut child_path = sub_root.add_node(sub_path, child)?;
            // We got a path from `sub_root` to the new child; update it to start at `self`.
            // FIXME: use push_front
```

    Child, 
    Foreign,
}
```

The state transition depends on both the node relation and the type of access operation.

```rust
enum AccessKind {
    Read, 
    Write,
}
```

Then we implement how to actually update and check states in a Tree during each memory access.

```rust
impl Node {
    /// Perform state transition on all locations of `self`.
    /// The `node_relation` indicates whether the accessed node is a child or foreign
    /// *from the perspective of `self`*.
    fn transition(
        &mut self, 
        node_relation: NodeRelation,
        access_kind: AccessKind,
        offset_in_alloc: Size,
        size: Size
    ) -> Result {
        let offset_start = offset_in_alloc.bytes();
        for offset in offset_start..offset_start + size.bytes() {
            self.location_states.mutate_at(offset, |location_state|{
                location_state.transition(access_kind, node_relation)
            })?;
        }

        ret(())
    }

    /// Recusively do state transition on `self` node and all its descendants.
    /// `path` is the path from `self` to the accessed node.
    /// `path` is None when the accessed node is not a descendant of `self`.

>     /// The `is_ancestor` indicates whether `self` is an ancestor of the accessed node.

Reviewer: This line is outdated

    ///
    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
    fn access(
        &mut self,
        path: Option<Path>, // self -> child
        access_kind: AccessKind,
        offset_in_alloc: Size,
        size: Size,
    ) -> Result {
        // Indicates whether the accessed node is a child or foreign
        // *from the perspective of `self`*.
        //
        // If `self` is an ancestor of the accessed node, the accessed node is a child of `self`.

>         // Otherwise, the accessed node is a `foreign` of `self`

Reviewer: ```suggestion
        // Otherwise, the accessed node is a `foreign` of `self`.
```

        let node_relation = if path.is_some() { NodeRelation::Child } else { NodeRelation::Foreign };

>         // perform state transition on `self`

Reviewer: ```suggestion
        // Perform state transition on `self`.
```

        self.transition(node_relation, access_kind, offset_in_alloc, size)?;

        for child_id in Int::ZERO..self.children.len() {
            // If the path starts with this child, do a child access with the path shortened by the first element.
            // Otherwise, do a child access with the path being None (i.e., child is not an ancestor of accessed node)
            let sub_path = if let Some(path) = path {
                match path.first() {

>                     Some(first) if first == child_id => Some(path.subslice_with_length(Int::ONE, path.len() - Int::ONE)),

Reviewer: This should also `split_first` once that exists:
`if let Some((head, tail)) = path.and_then(|p| p.split_first()) { ... }`
